File: backend/app/ai/agents.py
"""Hybrid boundary: use Gemini when a key is present, deterministic mock otherwise.

Gemini failures (network, schema) fall back to the mock so the pipeline never
hard-fails. Each function returns ``(result, used_ai)``.
"""
from __future__ import annotations

from typing import Callable, List, Optional, Tuple, TypeVar

from ..config import has_gemini
from ..logger import logger
from . import mock, prompts
from .gemini import generate_structured
from .schemas import (
    Classification,
    Extraction,
    Language,
    RecommendationResult,
    Risk,
    Validation,
    coerce_enum,
    DOCUMENT_TYPES,
    RECOMMENDATIONS,
    VALIDATION_STATUSES,
)

T = TypeVar("T")


def _with_fallback(label: str, real: Callable[[], T], fallback: Callable[[], T]) -> Tuple[T, bool]:
    if not has_gemini():
        return fallback(), False
    try:
        return real(), True
    except Exception as exc:  # noqa: BLE001
        logger.warn("agent.fallback", label=label, error=str(exc))
        return fallback(), False


def classify_document(text: str, images: Optional[List[str]] = None) -> Tuple[Classification, bool]:
    def real() -> Classification:
        result = generate_structured(Classification, prompts.classification_prompt(text), prompts.SYSTEM_BASE, images)
        result.document_type = coerce_enum(result.document_type, DOCUMENT_TYPES, "Unknown")
        return result

    return _with_fallback("classify", real, lambda: mock.mock_classify(text))


def detect_language(text: str) -> Tuple[Language, bool]:
    return _with_fallback(
        "language",
        lambda: generate_structured(Language, prompts.language_prompt(text), prompts.SYSTEM_BASE),
        lambda: mock.mock_language(text),
    )


def extract_fields(text: str, images: Optional[List[str]] = None) -> Tuple[Extraction, bool]:
    return _with_fallback(
        "extract",
        lambda: generate_structured(Extraction, prompts.extraction_prompt(text), prompts.SYSTEM_BASE, images),
        lambda: mock.mock_extract(text),
    )


def validate_document(document_type: str, fields: Extraction, ocr_confidence: float) -> Tuple[Validation, bool]:
    def real() -> Validation:
        result = generate_structured(
            Validation,
            prompts.validation_prompt(document_type, fields.model_dump(), ocr_confidence),
            prompts.SYSTEM_BASE,
        )
        result.validation_status = coerce_enum(result.validation_status, VALIDATION_STATUSES, "Needs Review")
        return result

    return _with_fallback("validate", real, lambda: mock.mock_validate(fields, ocr_confidence))


def assess_risk(
    document_type: str, fields: Extraction, ocr_confidence: float, validation: Validation
) -> Tuple[Risk, bool]:
    return _with_fallback(
        "risk",
        lambda: generate_structured(
            Risk,
            prompts.risk_prompt(document_type, fields.model_dump(), ocr_confidence, validation.reasons),
            prompts.SYSTEM_BASE,
        ),
        lambda: mock.mock_risk(validation, ocr_confidence),
    )


def recommend_decision(
    document_type: str,
    validation_status: str,
    risk_score: float,
    expired: bool,
    fields_complete: bool,
    reasons: List[str],
) -> Tuple[RecommendationResult, bool]:
    def real() -> RecommendationResult:
        result = generate_structured(
            RecommendationResult,
            prompts.recommendation_prompt(document_type, validation_status, risk_score, expired, fields_complete, reasons),
            prompts.SYSTEM_BASE,
        )
        result.recommendation = coerce_enum(result.recommendation, RECOMMENDATIONS, "Needs Human Review")
        return result

    return _with_fallback(
        "recommend",
        real,
        lambda: mock.mock_recommend(validation_status, risk_score, expired, fields_complete),
    )

File: lib/ai/agents.ts
import { generateStructured } from "./gemini";
import {
  ClassificationSchema,
  ExtractionSchema,
  LanguageSchema,
  RecommendationSchema,
  RiskSchema,
  ValidationSchema,
  type ClassificationResult,
  type ExtractionResult,
  type LanguageResult,
  type RecommendationResult,
  type RiskResult,
  type ValidationResultAI,
} from "./schemas";
import {
  SYSTEM_BASE,
  classificationPrompt,
  extractionPrompt,
  languagePrompt,
  recommendationPrompt,
  riskPrompt,
  validationPrompt,
} from "./prompts";
import { mockClassify, mockExtract, mockLanguage, mockRecommend, mockRisk, mockValidate } from "./mock";
import { hasGemini } from "../config";
import { logger } from "../logger";

// Each function is the "hybrid" boundary: it uses Gemini when a key is present
// and the deterministic heuristic otherwise. Gemini failures (network, schema)
// fall back to the mock so the pipeline never hard-fails.

async function withFallback<T>(label: string, real: () => Promise<T>, mock: () => T): Promise<{ value: T; usedAi: boolean }> {
  if (!hasGemini()) return { value: mock(), usedAi: false };
  try {
    return { value: await real(), usedAi: true };
  } catch (err) {
    logger.warn("agent.fallback", { label, error: (err as Error).message });
    return { value: mock(), usedAi: false };
  }
}

export function classifyDocument(text: string, images?: string[]) {
  return withFallback<ClassificationResult>(
    "classify",
    () =>
      generateStructured({
        system: SYSTEM_BASE,
        user: classificationPrompt(text),
        schema: ClassificationSchema,
        images,
      }),
    () => mockClassify(text)
  );
}

export function detectLanguage(text: string) {
  return withFallback<LanguageResult>(
    "language",
    () => generateStructured({ system: SYSTEM_BASE, user: languagePrompt(text), schema: LanguageSchema }),
    () => mockLanguage(text)
  );
}

export function extractFields(text: string, images?: string[]) {
  return withFallback<ExtractionResult>(
    "extract",
    () => generateStructured({ system: SYSTEM_BASE, user: extractionPrompt(text), schema: ExtractionSchema, images }),
    () => mockExtract(text)
  );
}

export function validateDocument(args: {
  documentType: string;
  fields: ExtractionResult;
  ocrConfidence: number;
}) {
  return withFallback<ValidationResultAI>(
    "validate",
    () =>
      generateStructured({
        system: SYSTEM_BASE,
        user: validationPrompt({ documentType: args.documentType, fields: args.fields, ocrConfidence: args.ocrConfidence }),
        schema: ValidationSchema,
      }),
    () => mockValidate(args.fields, args.ocrConfidence)
  );
}

export function assessRisk(args: {
  documentType: string;
  fields: ExtractionResult;
  ocrConfidence: number;
  validation: ValidationResultAI;
  images?: string[];
}) {
  return withFallback<RiskResult>(
    "risk",
    () =>
      generateStructured({
        system: SYSTEM_BASE,
        user: riskPrompt({
          documentType: args.documentType,
          fields: args.fields,
          ocrConfidence: args.ocrConfidence,
          validationReasons: args.validation.reasons,
        }),
        schema: RiskSchema,
        images: args.images,
      }),
    () => mockRisk(args.validation, args.ocrConfidence)
  );
}

export function recommendDecision(args: {
  documentType: string;
  validationStatus: string;
  riskScore: number;
  expired: boolean;
  fieldsComplete: boolean;
  reasons: string[];
}) {
  return withFallback<RecommendationResult>(
    "recommend",
    () =>
      generateStructured({
        system: SYSTEM_BASE,
        user: recommendationPrompt(args),
        schema: RecommendationSchema,
      }),
    () =>
      mockRecommend({
        validationStatus: args.validationStatus,
        riskScore: args.riskScore,
        expired: args.expired,
        fieldsComplete: args.fieldsComplete,
      })
  );
}
